crates/shared/src/price_estimation/native/oneinch.rs

```diff
@@ -0,0 +1,140 @@
+use {
+    super::{NativePriceEstimateResult, NativePriceEstimating},
+    crate::price_estimation::PriceEstimationError,
+    anyhow::{anyhow, Result},
+    ethrpc::current_block::{into_stream, CurrentBlockStream},
+    futures::{future::BoxFuture, FutureExt, StreamExt},
+    primitive_types::{H160, U256},
+    reqwest::Client,
+    std::{
+        collections::HashMap,
+        sync::{Arc, Mutex},
+    },
+};
+
+const BASE_URL: &str = "https://api.1inch.dev/";
+
+struct OneInch {
+    // Denominated in wei
+    prices: Arc<Mutex<HashMap<H160, U256>>>,
+}
+
+impl OneInch {
+    #[allow(dead_code)]
+    pub fn new(client: Client, chain_id: u64, current_block: CurrentBlockStream) -> Self {
+        let instance = Self {
+            prices: Arc::new(Mutex::new(HashMap::new())),
+        };
+        instance.update_prices_in_background(client, chain_id, current_block);
+        instance
+    }
+
+    fn update_prices_in_background(
+        &self,
+        client: Client,
+        chain_id: u64,
+        current_block: CurrentBlockStream,
+    ) {
+        let prices = self.prices.clone();
+        tokio::task::spawn(async move {
+            let mut block_stream = into_stream(current_block);
+            loop {
+                match update_prices(&client, chain_id).await {
+                    Ok(new_prices) => {
+                        tracing::debug!("OneInch spot prices updated");
+                        *prices.lock().unwrap() = new_prices;
+                    }
+                    Err(err) => {
+                        tracing::warn!(?err, "OneInch spot price update failed");
+                    }
+                }
+                block_stream.next().await;
```

Comment:
Have you tried to implement background work using `Maintaining` trait?
Seems to fit nicely especially since the `current_block` is only used to indicate when the update needs to be done.

Reply:
This is definitely possible. The problem with that is that it makes wiring up the price estimator harder. Price estimators are created in `price_estimation/factory` and maintainers are wired up in the top level `main.rs`. We could push a mutable reference to the `maintainers` vec into the factory and make the factory method have side effects, but I thought simply passing down this stream is simpler.
I agree though there is an inconsistency and we should either kill the `Maintaining` trait and make all maintainers manage their own task or replace the other uses of current_block_stream such as the account [balance cache](https://github.com/cowprotocol/services/blob/9bcf411bf87a4e11f515d6969095db61dd832847/crates/shared/src/account_balances/cached.rs#L117).

Reply:
I think spawning a background task is fine. In fact, we discovered along the way that the `Maintaining` abstraction wasn't great (and might make sense to re-write existing `Maintaining` implementations to also spawn background tasks for updating on new blocks).

I don't remember the exact time in the past where we started moving away from using the `Maintaining` trait, but I think @MartinquaXD and I discussed changing this in the context of "getting rid of the `shared` crate.